### PR TITLE
Backport/william/v4.2.12

### PR DESCRIPTION
**Pull request by qzhuyan**
backport fixes to 4.2 for the following issues:

#2671
#2985

**If your build fails** due to your commit message not passing the build checks, please review the guidelines here: https://github.com/emqx/emqx/blob/master/CONTRIBUTING.md.

## PR Checklist
Please convert it to a draft if any of the following conditions are not met. Reviewers may skip over until all the items are checked:

- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)
- [ ] In case of non-backward compatible changes, reviewer should check this item as a write-off, and add details in **Backward Compatibility** section

## More information

### CODE REVIEW

**Review by k32 — DISMISSED** [review]

File: priv/emqx.schema [hunk 1]
```diff
@@ -2105,6 +2105,18 @@ end}.
   {datatype, flag}
 ]}.
 
+%% @doc Performance toggle for subscribe/unsubscribe wildcard topic.
+%%      Change this toggle only when there are many wildcard topics.
+%% key:   mnesia translational updates with per-key locks. recommended for single node setup.
+%% tab:   mnesia translational updates with table lock. recommended for multi-nodes setup.
+%% global: global lock protected updates. recommended for larger cluster.
+%% NOTE: when changing from/to 'global' lock, it requires all nodes in the cluster
```
Comment (zmstone) [inline priv/emqx.schema:2113]:
```suggestion
%% NOTE: when changing from/to 'global' lock, it requires a full cluster restart.
```

Comment (zmstone) [reply priv/emqx.schema:2113]: Sorry. I forgot to put 'restart' in the comment.

File: src/emqx.appup.src [hunk 2]
```diff
@@ -8,23 +8,49 @@
     end,
 {VSN,
   [
+   {"4.2.10", [
```
Comment (zmstone) [inline src/emqx.appup.src:11]: 4.2.11 ?

File: src/emqx_router.erl [hunk 3]
```diff
@@ -247,10 +249,59 @@ delete_trie_route(Route = #route{topic = Topic}) ->
     end.
 
 %% @private
+-spec(maybe_trans(function(), list(any())) -> ok | {error, term()}).
+maybe_trans(Fun, Args) ->
+    case persistent_term:get(emqx_route_lock_type) of
```
Comment (zmstone) [inline src/emqx_router.erl:254]: hot beam reload will not have the PT initialised (init code is in `emqx_router_sup`)
so we need a default value here.

Comment (qzhuyan) [reply src/emqx_router.erl:254]: right, I should cherry pick from e4.2.6

**Review by zmstone — DISMISSED** [review]

**Review by zmstone — APPROVED** [review]

**Pull request merged**